solutions/ys-MaxIndepSet.mdx

```diff
@@ -0,0 +1,207 @@
+---
+id: ys-MaxIndepSet
+source: YS
+title: Max Indep Set
+author: Shijie Ren
+---
+
+<Spoiler title="Hint 1">
+
+Generate the complement of the given graph.
+
+The complement or inverse of a graph G is a graph H on the same vertices
+such that two distinct vertices of H are adjacent if and only if
+they are **not** adjacent in G.
+
+That is, to generate the complement of a graph,
+one fills in all the missing edges required to form a complete graph,
+and removes all the edges that were previously there.
+
+</Spoiler>
+
+<Spoiler title="Hint 2">
+
+Consider Bitmask DP when $ N \leq 20 $.
+
+</Spoiler>
+
+<Spoiler title="Hint 3">
+
+If you stuck on DP when $ N \leq 20 $,
+
+Let the DP array store **the set of points** in the subgraph
+that meet specific conditions.
+
+</Spoiler>
+
+<Spoiler title="Hint 4">
+
+Divide the original graph into two parts, do DP respectively, and then merge.
+
+</Spoiler>
+
+<Spoiler title="Explanation">
+
+#### Bitmask DP
+
+First we can turn the original problem into another problem
+by building the complement graph of the original graph.
+
+Then we can turn it into the maximum clique problem (MCP).
+
+Clique can be seen as a complete graph.
+
+In fact, it is not a good way to store only the number of points in the answer graph ,
+because then it will be hard to transfer.
+
+Then switch ideas.
+
+Let the DP array store the point set of the maximum clique in
+the subgraph.
+
+That is, the element in DP array is also a bitmask.
+
+Simply use $\texttt{dp}[S]$ to represent **any** valid maximum clique solution in
+subgraph $S$.
+
+Consider a point $k$ in $S$.
+
+if this point is in the maximum clique,
+
+$$
+\texttt{dp}[S] = \texttt{dp}[E_k\cap S] \cup \{k\}
+$$
+
+(where $E_k$ represents the neighbour set of $k$)
+
+otherwise
+
+$$
+\texttt{dp}[S] =\texttt{dp} [S / {k}]
+
+$$
+
+Take the maximum one.
+
+$$
+\texttt{dp}[S] = \max\{ \texttt{dp}[E_k\cap S] \cup \{k\} , \texttt{dp} [S / {k}] \}
+$$
+
+**Notice this** $\max$ **compared by popcount in the program.**
+
+But the time complexity is $\mathcal{O}(2 ^ n)$ or $\mathcal{O}(2 ^ n \cdot n)$,
+which needs further optimization.
+
+#### Meet-in-the-middle
+
+Divide the original graph into two parts, do DP respectively, and then merge:
+
+if the two parts are $S1$ and $S2$, loop **every subgraph** of $S1$,
+
+find subgraph of $S2$ that have direct edges with **all points** in subgraph of
+$S1$ that we loop,
+
+do set UNION operation with two subgraph in $S1$ and $S2$.
+
+"Subgraph of $S2$ that have direct edges with **all points** in subgraph of $S1$
+that we loop" can be solved with another simple Bitmask DP.
+
+Time Complexity is $\mathcal{O}(2^{\frac N2})$.
+
+</Spoiler>
+
+<Spoiler title="Implementation">
+
+**Time Complexity:** $\mathcal{O}(2^{\frac N2})$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <cstdio>
+using ll = long long;
+const int N = 45;
```

Comment:
isn't bitwise for N = 20?
just use `const int N = 20` and then when the need arises for 40 use `N * 2` instead